[server] 데이터 날짜 필터링 기능

bohyeon-n opened the pull request:
- 예약 날짜 필터링 기능 추가함 
- pagination할 때 limit과 offset이 잘못되어 있어서 수정함 
- 숙소 조회할 때 쿼리문에 pagination이 없어서 추가해줌

일단 예약일 필터링 먼저 구현했습니다. 보시고 알려주세요!
예약 작업은 내일 하도록 하겠습니다.

bohyeon-n committed:
[#19]feature: 숙소 조회 예약일 필터링 추가

이미 예약된 숙소를 제외한 나머지 숙소를 가져오기 위해서 예약일에 대한 필터링 추가함

bohyeon-n committed:
[#19]fix: pagination offset, limit 값 수정함

limit은 한 페이지에 들어가는 데이터 개수, offset은 시작점 값을 넣어주어야 하는데
잘못 넣어서 수정함

bohyeon-n committed:
[#19]refactor: 불필요한 import문 삭제함

kses1010 commented on BE/src/main/java/com/codesquad/airbnb/controller/RoomController.java:
```diff
@@ -32,8 +32,8 @@
                                             @RequestParam(value = "maximum-price", required = false, defaultValue = "0") int maxPrice) {
-        int offset = page - 1;
-        int limit = page * 9;
+        int limit = 9;
```
limit변수는 변하지 않으니 limit를 `private final` 로 해요!!

kses1010 commented on BE/src/main/java/com/codesquad/airbnb/repository/RoomDao.java:
```diff
@@ -28,19 +28,26 @@ public RoomDao(DataSource dataSource) {
         String roomsSql = "SELECT id, title, thumbnail, super_host, address, location, accommodates, " +
                 "price, cleaning_fee, review_score " +
                 "from room " +
-                "WHERE accommodates >= :totalGuest " +
-                "and price >= :minPrice ";
+                "WHERE id not in ( select distinct (r.room_id) " +
+                "from reservation_date rd left join reservation r on rd.reservation_id = r.id where rd.reservation_date between :checkIn and :checkOut) " +
+                "and room.accommodates >= :totalGuest " +
+                "and room.price >= :minPrice ";
 
         if (maxPrice != 0) {
-            roomsSql += "and price <= :maxPrice";
+            roomsSql += "and price <= :maxPrice ";
         }
 
+        roomsSql += "limit :limit offset :offset";
```
SQL문의 예약어들은 대문자로 지정해서 SQL 컨벤션을 맞추면 될거 같아요.

그리고 너무 길어진다면 예약어 기준으로 줄바꿈하여 코드 보기 쉽게 부탁드려요. 인텔리 기준으로 실선넘어가면 개행하는걸로해주세욥.

Between 문이  10 <= A < 12 이런 느낌이라 찾아보니 
`WHERE rd.reservation_date >= ? AND rd.reservation_date <= ?` 
이런식으로 넣으면 설정하면 저희가 원하는 날짜 검색이 될 듯 합니다.

kses1010 replied:
이 조건문으로 검색하면 예약되어 있는 방이 나오더라구요. 
그래서 `left join reservation r on rd.reservation_id != r.id` 이렇게 하면 예약 되어있지 않은 방이 나옵니다.

MySQL에서 `SELECT DATEDIFF('2020-05-30', '2020-05-01' - interval 1 DAY);` 이런 식으로 하면 날짜 뺄셈이 가능합니다. 이걸로 예약이나 totalPrice 계산에 도움이 됩니다. 참고해주세요~

bohyeon-n replied:
@kses1010 리뷰 감사합니다~
`rd.reservation_id != r.id` 이렇게 하면 reservation_date에서 일치 하지 않는 reservation_id 를 가진 데이터를 찾을 것 같은데 
혹시 어떤 상황에서 예약된 방이 나왔는지 알 수 있을까요?

kses1010 replied:
아 잘못 봤네요 NOT IN이 대신하는군요 착각했습니다. BETWEEN만 바꾸시면 될 듯 합니다!

bohyeon-n commented:
나머지는 예약 기능 만들면서 수정해볼게요 ~

bohyeon-n committed:
[#26]feature: ReservationForm 객체 추가

예약 api에서 RequestBody로 받을 reservationForm 객체 추가함

bohyeon-n committed:
[#26]feature: Date 차를 계산하는 DayCalculator 클래스 추가함

checkIn과 checkOut 사이의 일수를 개산하기 위해서 getDiffDays 메소드를 작성함

bohyeon-n committed:
[#26]feature: room 예약 기능 구현

reservation에 데이터 추가함
checkIn과 checkOut의 차이만큼 reservation_date 를 추가해줌
checkOut은 나가는 날이므로 checkOut 날짜는 추가하지 않음

bohyeon-n committed:
[#26]fix: reservation_date 추가 기능 수정

reservation_date에 추가할 떄 reservation_id를 넣도록 수정함
insert 수행할 때 reservation_id를 받아오도록 했음
서비스와 관련된 로직은 RoomService에 있는 것이 적합할 것 같아 이동함

bohyeon-n committed:
[#26]fix: limit, offset 상수값 final로 수정함

bohyeon-n committed:
[#26]refactor: 긴 sql문 코드 정리함

bohyeon-n commented:
@kses1010 예약 하기 기능 중간 공유합니다! 

추가할 기능 
- [x] 예약 요청 데이터가 유효한지 체크하기 
- [ ] ~~에러 처리하기~~
- [ ] ~~userId 추가하기 (oAuth한 다음에)~~
- [ ] 적절한 응답값 주기

kses1010 commented:
> @kses1010 예약 하기 기능 중간 공유합니다!
> 
> 추가할 기능
> 
> * 예약 요청 데이터가 유효한지 체크하기
> * 에러 처리하기
> * userId 추가하기 (oAuth한 다음에)
> * 적절한 응답값 주기

kses1010 commented:
지금은 예약할 때 하드코딩된 유저로 하실 수 있나요?

bohyeon-n committed:
[#26]refactor: SQL문의 예약어들 대문자로 수정함

sql 컨벤션을 맞추기 위해서 예약어들은 대문자로 수정함

bohyeon-n committed:
[#26]feature: resrevationForm 값이 유효한지 체크하기 기능 추가

checkIn, checkOut 값이 유효한 값인지 체크하는 로직 추가함
적절한 값이 들어오지 않았을 때 예외 처리 필요함

bohyeon-n committed:
[#26]feature: 예약 성공 응답 추가

예약이 성공했을 때 OK 응답함

bohyeon-n commented:
@kses1010
같이 구현하기로 한 부분을 제외하고 예약하기 기능을 구현하였습니다. 
between은 checkOut 마지막 날은 조회하지 않으므로 적절한 것 같아서 수정하지 않았습니다! 
응답값에 대한 부분은 간단하게 했는데 내일 같이 얘기하면 좋을 것 같아요!

kses1010 committed:
[#26] feature: 조건에 따른 총 가격 기능 추가

- Guest, Price, ReservationDate는 사용하지 않아서 삭제
- RoomMapper
  - totalPrice(): 슈퍼호스트 존재여부에 따른 총가격 표시

bohyeon-n committed:
[#26]feature: 데이터베이스 조회해서 받아오는 데이터와 응답 데이터를 구분함

데이터 베이스를 조회한 것과 응답을 위한 데이터를 구분하기 위해서 Room, RoomResponse dto를 만듦

bohyeon-n committed:
[#26]fix: webConfig 파일 수정

WebMvcConfigure implements하도록 수정함

bohyeon-n committed:
[#26]feature: Room, RoomResponse dto 클래스 추가

데이터 가져오는 Room과 응답 객체를 따로 작성하여 totalPrice, discountPrice는  RoomResponse에서 넣어줌

bohyeon-n committed:
[#26]refactor: RoomRepository 삭제

사용하지 않는 RoomRepository 삭제함

bohyeon-n committed:
[#26]feature: 숙소 조회 응답값을 수정한 것을 적용함

숙소 조회 응답값을 RoomResponse 객체로 수정한 것을 적용함
